### PR TITLE
Keep rules marked promote until-clean with -p

### LINKED ISSUE #4401: Bad interaction between '(mode promote-until-clean)' and tests ran with '-p <pkg>'
## Expected Behavior

When using a file (let's say `test`) that is "promoted until clean" as a dep for a test defined for a specific package `foo`, `dune build @runtest -p foo` should succeed.

## Actual Behavior

`dune build @runtest -p foo` fails with `Error: No rule found for test`.
The conjunction of `(mode promote-until-clean)` and `-p <pkg>` seems to make dune "forget" about how to build some targets. I don't know if that is expected and if so, what is the correct way to do things.

## Reproduction

See https://gist.github.com/Gbury/598f0b6f751767e8a66fe9c508704976

dune file:
```
(rule
  (target test)
  (mode promote-until-clean) ; without this, everything runs fine 
                             ; but with it, dune complains about "No rule found for test"
  (action (with-stdout-to %{target} (run echo foobar)))
)

(rule
  (alias runtest)
  (package foo)
  (deps (:test test))
  (action (diff %{test} reference))
)
```

0. `dune build @runtest` succeeds as can be expected
1. `dune build @runtest -p foo` fails with "Error: No rule found for test"  (make sure to `dune clean` between attempts, particularly after a successful run)
2. If the `(mode promote-until-clean)` is commented out, `dune build @runtest -p foo` succeeds

## Specifications

- Version of `dune` (output of `dune --version`): 2.8.4
- Version of `ocaml` (output of `ocamlc --version`): 4.11.1
- Operating system (distribution and version): archlinux 5.11.9-arch1-1

## Additional information 

- Link to gist with verbose output (run `dune` with the `--verbose` flag):
Output of `dune build @runtest -p foo --verbose` :
https://gist.github.com/Gbury/80ed0220ccdd729781b35d445731de90

In practice, this happens to me in the dolmen project (see e.g. [this dune file](https://github.com/Gbury/dolmen/blob/687ae8b46b0e7df3f414054b31e52a78e9185b31/src/languages/ae/dune#L34-L66 )), where a rule is defined to generate an updated version of menhir's error message file (which is promoted-until-clean, since it is meant to be used as a reference when updating the error messages after changing the grammar), and another rule is defined to use that file to check that all error states have an error message written.

### PR DESCRIPTION
**Pull request by emillon**
Closes #4401

### CODE REVIEW

File: src/dune_rules/dune_file.ml [hunk 1]
```diff
@@ -2332,15 +2332,27 @@ type t =
   ; stanzas : Stanzas.t
   }
 
+let is_promoted_rule version rule =
+  let is_promoted_mode = function
+    | Rule.Mode.Promote { only = None; lifetime; _ } ->
+      if version >= (3, 4) then
```
Comment (rgrinberg) [inline src/dune_rules/dune_file.ml:2338]: `version >= (3, 4) || (match lifetime with Unlimited -> true | Until_clean -> false)`

**Review by rgrinberg — APPROVED** [review]

File: src/dune_rules/dune_file.ml (hunk 1 above)
Comment (emillon) [reply src/dune_rules/dune_file.ml:2338]: That would be `not (version >= ...) || match ...` so at that point I think that the explicit `if` is clearer.

File: test/blackbox-tests/test-cases/github4401.t [hunk 2]
```diff
@@ -21,3 +21,12 @@ are ignored. See #4401.
   Error: No rule found for test
   -> required by alias runtest in dune:5
   [1]
+
+This is correctly ignored if dune-lang is bumped to 3.5.
```
Comment (christinerose) [inline test/blackbox-tests/test-cases/github4401.t:25]:
```suggestion
This is correctly ignored if `dune-lang` is bumped to 3.5.
```

**Review by christinerose — APPROVED** [review]
One suggestion to wrap `dune-lang` in monospace

File: test/blackbox-tests/test-cases/github4401.t (hunk 2 above)
Comment (Alizter) [reply test/blackbox-tests/test-cases/github4401.t:25]: @christinerose This is inside a cram test however and not in any documentation. So monospace is really up to the font you are using 0:-)

**Pull request merged**